src/Microsoft.ML.Transforms/NAIndicatorTransform.cs

```diff
@@ -68,116 +78,196 @@ private static VersionInfo GetVersionInfo()
+            {
+                schema.TryGetColumnIndex(ColumnPairs[i].input, out int colSrc);
+                var type = schema.GetColumnType(colSrc);
+
+                if (!type.IsVector)
+                {
+                    inputTypes[i] = type.AsPrimitive;
+                    outputTypes[i] = BoolType.Instance;
+                }
+                else
+                {
+                    inputTypes[i] = new VectorType(type.ItemType.AsPrimitive, type.AsVector);
+                    outputTypes[i] = new VectorType(BoolType.Instance, type.AsVector);
+                }
+            }
+            return (inputTypes, outputTypes);
+        }
 
         /// <summary>
         /// Convenience constructor for public facing API.
         /// </summary>
         /// <param name="env">Host Environment.</param>
         /// <param name="input">Input <see cref="IDataView"/>. This is the output from previous transform or loader.</param>
-        /// <param name="name">Name of the output column.</param>
-        /// <param name="source">Name of the column to be transformed. If this is null '<paramref name="name"/>' will be used.</param>
-        public NAIndicatorTransform(IHostEnvironment env, IDataView input, string name, string source = null)
-            : this(env, new Arguments() { Column = new[] { new Column() { Source = source ?? name, Name = name } } }, input)
+        /// <param name="columns">Specifies the names of the input columns for the transform and the resulting output column names.</param>
+        public NAIndicatorTransform(IHostEnvironment env, IDataView input, params ColumnInfo[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(NAIndicatorTransform)), GetColumnPairs(columns))
         {
+            // Check that all the input columns are present and correct.
+            for (int i = 0; i < ColumnPairs.Length; i++)
+            {
+                if (!input.Schema.TryGetColumnIndex(ColumnPairs[i].input, out int srcCol))
+                    throw Host.ExceptSchemaMismatch(nameof(input), "input", ColumnPairs[i].input);
+                CheckInputColumn(input.Schema, i, srcCol);
+            }
+            (_inputTypes, _outputTypes) = GetTypes(input.Schema);
+        }
+
+        private NAIndicatorTransform(IHost host, ModelLoadContext ctx)
+            : base(host, ctx)
+        {
+            Host.AssertValue(ctx);
+            _outputTypes = null;
+            _inputTypes = null;
         }
 
         /// <summary>
-        /// Public constructor corresponding to SignatureDataTransform.
+        /// Factory method for SignatureDataTransform.
         /// </summary>
-        public NAIndicatorTransform(IHostEnvironment env, Arguments args, IDataView input)
-            : base(env, RegistrationName, Contracts.CheckRef(args, nameof(args)).Column,
-                input, TestType)
+        public static IDataTransform Create(IHostEnvironment env, Arguments args, IDataView input)
```

Comment:
>public [](start = 8, length = 6)

can it be internal?

Reply:
even private!

---
In reply to: [224316542](https://github.com/dotnet/machinelearning/pull/1217#discussion_r224316542) [](ancestors = 224316542)